src/cowboy_telemetry_h.erl

```diff
@@ -7,138 +7,51 @@
-terminate(StreamID, {ErrorType, _, _} = Reason, #state{next=Next, start_time=StartTime})
-    when ErrorType == socket_error;
-         ErrorType == connection_error ->
-    emit_stop_error_event(StreamID, StartTime, Reason),
-    cowboy_stream:terminate(StreamID, Reason, Next);
-terminate(StreamID, Reason, #state{next=Next}) ->
-    cowboy_stream:terminate(StreamID, Reason, Next).
+terminate(StreamID, Reason, State) ->
+    cowboy_metrics_h:terminate(StreamID, Reason, State).
 
-early_error(StreamID, Reason, PartialReq, Resp0, Opts) ->
-    Resp = cowboy_stream:early_error(StreamID, Reason, PartialReq, Resp0, Opts),
-    emit_early_error_event(StreamID, Reason, PartialReq, Resp),
-    Resp.
+early_error(StreamID, Reason, PartialReq, Resp, Opts) ->
+    cowboy_metrics_h:early_error(StreamID, Reason, PartialReq, Resp, add_metrics_callback(Opts)).
 
 %
 
-fold([], State, _Acc) ->
-    State;
-
-fold([{response, _, _, _} = Response | Tail],
-     #state{streamid=StreamID, start_time=StartTime} = State,
-     Acc) ->
-    emit_stop_event(StreamID, StartTime, Response),
-    fold(Tail, State#state{emit=done}, Acc);
-
-fold([{data, fin, _} | Tail],
-     #state{streamid=StreamID, start_time=StartTime, chunked_resp_status=RespStatus, chunked_resp_headers=RespHeaders} = State,
-     Acc) ->
-    emit_stop_event(StreamID, StartTime, {response, RespStatus, RespHeaders, nil}),
-    fold(Tail, State#state{emit=done, chunked_resp_status=undefined, chunked_resp_headers = undefined}, Acc);
+add_metrics_callback(Opts) ->
+    maps:put(metrics_callback, fun metrics_callback/1, Opts).
 
-fold([{internal_error, {'EXIT', _, Reason}, _} | Tail],
-     #state{streamid=StreamID, start_time=StartTime} = State,
-     #acc{error_response=ErrorResponse}) ->
-    emit_exception_event(StreamID, StartTime, Reason, ErrorResponse),
-    fold(Tail, State#state{emit=done}, #acc{});
-
-fold([{headers, RespStatus, RespHeaders} | Tail], State, Acc) ->
-    fold(Tail, State#state{chunked_resp_status=RespStatus, chunked_resp_headers=RespHeaders}, Acc);
-
-fold([{error_response, _, _, _} = ErrorResponse | Tail], State, Acc) ->
-    fold(Tail, State, Acc#acc{error_response=ErrorResponse});
-
-fold([_ | Tail], State, Acc) ->
-    fold(Tail, State, Acc).
-
-emit_start_event(StreamID, SystemTime, Req) ->
+metrics_callback(#{reason := {internal_error, {'EXIT', _, {_, Stacktrace}}, _}, req_start := ReqStart, req_end := ReqEnd} = Metadata) ->
     telemetry:execute(
-        [cowboy, request, start],
-        #{system_time => SystemTime},
-        #{stream_id => StreamID, req => Req}
-    ).
-
-emit_stop_event(StreamID, StartTime, Response) ->
-    EndTime = erlang:monotonic_time(),
+        [cowboy, request, exception],
+        #{duration => ReqEnd - ReqStart},
+        maps:merge(Metadata, #{kind => exit, stacktrace => Stacktrace}));
+metrics_callback(#{reason := {ErrorType, _, _} = Reason, req_start := ReqStart, req_end := ReqEnd} = Metadata)
```

Comment:
The only reason this function head is here is to match the `span` "spec" by adding `error => Reason` to the `metadata`, otherwise it could be handled by the function head below